nsxt/resource_nsxt_principal_identity.go

```diff
@@ -0,0 +1,179 @@
+		mpRolesForPath[i].DeletePath = pRolesForPath.DeletePath
+		mpRolesForPath[i].Path = pRolesForPath.Path
+		mpRolesForPath[i].Roles = make([]mpModel.Role, len(pRolesForPath.Roles))
+		for j, pRole := range pRolesForPath.Roles {
+			mpRolesForPath[i].Roles[j].Role = pRole.Role
+			mpRolesForPath[i].Roles[j].RoleDisplayName = pRole.RoleDisplayName
+		}
+	}
+	return mpRolesForPath
+}
+
+func convertToPolicyRolesForPath(mpRolesForPath []mpModel.RolesForPath) []policyModel.RolesForPath {
+	pRolesForPath := make([]policyModel.RolesForPath, len(mpRolesForPath))
+	for i, mRolesForPath := range mpRolesForPath {
+		pRolesForPath[i].DeletePath = mRolesForPath.DeletePath
+		pRolesForPath[i].Path = mRolesForPath.Path
+		pRolesForPath[i].Roles = make([]policyModel.Role, len(mRolesForPath.Roles))
+		for j, mRole := range mRolesForPath.Roles {
+			pRolesForPath[i].Roles[j].Role = mRole.Role
+			pRolesForPath[i].Roles[j].RoleDisplayName = mRole.RoleDisplayName
+		}
+	}
+	return pRolesForPath
+}
+
+func resourceNsxtPrincipleIdentityCreate(d *schema.ResourceData, m interface{}) error {
+	connector := getPolicyConnector(m)
+	client := principal_identities.NewWithCertificateClient(connector)
+
+	tags := getMPTagsFromSchema(d)
+	name := d.Get("name").(string)
+	nodeID := d.Get("node_id").(string)
+	certificatePem := d.Get("certificate_pem").(string)
+	rolesForPaths := convertToMPRolesForPath(getRolesForPathList(d))
+
+	piObj := mpModel.PrincipalIdentityWithCertificate{
+		Tags:           tags,
+		Name:           &name,
+		NodeId:         &nodeID,
+		CertificatePem: &certificatePem,
+		RolesForPaths:  rolesForPaths,
+	}
+
+	pi, err := client.Create(piObj)
+	if err != nil {
+		return handleCreateError("PrincipalIdentity", name, err)
+	}
+	d.SetId(*pi.Id)
+
+	return resourceNsxtPrincipleIdentityRead(d, m)
+}
+
+func resourceNsxtPrincipleIdentityRead(d *schema.ResourceData, m interface{}) error {
+	connector := getPolicyConnector(m)
+	client := trust_management.NewPrincipalIdentitiesClient(connector)
+	id := d.Id()
+	if id == "" {
+		return fmt.Errorf("error obtaining logical object id")
+	}
+	piObj, err := client.Get(id)
+	if err != nil {
+		return handleReadError(d, "PrincipalIdentity", id, err)
+	}
+
+	setMPTagsInSchema(d, piObj.Tags)
+	d.Set("name", piObj.Name)
+	d.Set("node_id", piObj.NodeId)
+	d.Set("certificate_id", piObj.CertificateId)
+	setRolesForPathInSchema(d, convertToPolicyRolesForPath(piObj.RolesForPaths))
+	return nil
+}
+
+func resourceNsxtPrincipleIdentityDelete(d *schema.ResourceData, m interface{}) error {
+	connector := getPolicyConnector(m)
+	piClient := trust_management.NewPrincipalIdentitiesClient(connector)
+	id := d.Id()
+	if id == "" {
+		return fmt.Errorf("error obtaining logical object id")
+	}
+	certID := d.Get("certificate_id").(string)
+	err := piClient.Delete(id)
+	if err != nil {
+		return handleDeleteError("PrincipalIdentity", id, err)
+	}
+
+	// Clean up underlying cert imported by NSX if exists
+	if len(certID) > 0 {
```

Comment:
sounds like the cert is created automatically by the create API, but not deleted nu delete API?

Reply:
Right.